src/plugin.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+struct EventKey {
+  byte row_;
+  byte col_;
+  uint8_t keyState_;
+};
+
+class KaleidoscopePlugin {
+
+  friend class Kaleidoscope_;
+
+ protected:
+
+  /** Initial plugin setup hook.
+   * All plugins are supposed to provide a singleton object, statically
+   * initialized at compile-time (with few exceptions). Because of this, the
+   * order in which they are instantiated is unspecified, and cannot be relied
+   * upon. For this reason, one's expected to explicitly initialize, "use" the
+   * plugins one wishes to, by calling `Kaleidoscope.use()` with a list of plugin
+   * object pointers.
+   *
+   * This function will in turn call the `begin` function of each plugin,
+   * so that they can perform any initial setup they wish, such as registering
+   * event handler or loop hooks. This is the only time this function will be
+   * called. It is intentionally protected, and accessible by the `Kaleidoscope`
+   * class only.
+   */
+  virtual void begin(void)
+  __attribute__((deprecated("KaleidoscopePlugin.begin() is deprecated. Please implement KaleidoscopePlugin.init() instead."))) {
+  };
+
+ public:
+
+  // The following callbacks handle the synchronized communication
+  // between the Kaleidoscope core and its plugins.
+  //
+  // A plugin can, but does not have to implement all provided hook methods.
+  // Calls to unimplemented hook methods are entirely optimized away by the compiler.
+  // Thus, a non-implemented hook causes neither memory consumption nor
+  // run-time overhead.
+  //
+  // Note: All hook methods in this class are non virtual on purpose as the actual
+  //       interface between the Kaleidoscope singleton class and the
+  //       plugins is the polymorphic PluginHookAdapter class
+
+  void init() {
+    // By letting the new init method call the legacy
+    // begin() method, we make sure that the old hooking
+    // interface is supported even if KALEIDOSCOPE_INIT_PLUGINS()
+    // and KALEIDOSCOPE_CONNECT_PLUGINS are used to register
+    // a plugin that relies on the legacy begin() method
+    // to initialize itself and register hooks.
+    //
+    this->begin();
+  }
+
+  // This handler is supposed to return false if no other handlers are
+  // supposed to be called after it, false otherwise.
+  //
+  // The handler is allowed to modify the mappedKey that is therefore
+  // passed by reference.
+  //
+  bool eventHandlerHook(Key &mappedKey, const EventKey &eventKey) {
+    return true; // Always allow other handlers to continue
```

Comment:
Does this `return true` prevent the compiler from optimizing out the function call entirely for a plugin that doesn't (re-)implement the `eventHandlerHook()` method? Obviously, we want it here, regardless, but I'm wondering if there's a cost to other hooks being added that behave the same way.

Reply:
The function is inline and the return value a compile time constant. This enables the compiler to optimize out the entire function. The value is then processed by the inlined predicated and used in an if statement. Everything is inline and constant. This even leads to all other plugin calls being possibly eliminated once the first plugin returns false (dead code).

Reply:
I'm still not sure what the answer to my question is, so I'll be more specific. If there are twelve plugin classes that don't override this hook, does the process end up performing more instructions than if there are only three such plugins?

Reply:
Ok, let's see. To demonstrate what happens, I did the following. I added extra entries of plugins that do not (yet) support the new plugin interface to the list of plugins registered with the new interface. The purpose of this is to simulate plugins that do not implement one or more of some hooks.
```cpp
...
KALEIDOSCOPE_INIT_PLUGINS(

  // The MouseKeys plugin lets you add keys to your keymap which move the mouse.
  MouseKeys,
  Macros,
  Macros
)
...
```
When we look at the disassembly (Kaleidoscope calls this a *decompile* - how I would love a decompiler ;-) ), we see the following.

```
000005e0 <PluginHookAdapter<HookLoop>::postReportHook()>:
     5e0:       10 92 17 06     sts     0x0617, r1      ; 0x800617 <MouseKeys_::mouseMoveIntent>
     5e4:       08 95           ret
```
The calls to the `Macro`s' inlined empty `postReportHook`s (those of the base class, actually) vanish.
```cpp
000005f4 <PluginHookAdapter<HookLoop>::preReportHook()>:
     5f4:       89 e0           ldi     r24, 0x09       ; 9
     5f6:       96 e0           ldi     r25, 0x06       ; 6
     5f8:       0c 94 07 10     jmp     0x200e  ; 0x200e <MouseKeys_::preReportHook()>
```
So do the calls to their `preReportHook`s...
```cpp
000005fc <PluginHookAdapter<HookLoop>::eventHandlerHook(Key_&, EventKey const&)>:
     5fc:       89 e0           ldi     r24, 0x09       ; 9
     5fe:       96 e0           ldi     r25, 0x06       ; 6
     600:       0c 94 84 10     jmp     0x2108  ; 0x2108 <MouseKeys_::eventHandlerHook(Key_&, EventKey const&)>
```
...and the `eventHandlerHook`s.

The only calls that persist are those to the Macro pluginsa' init methods. Those cannot be optimized out, as base class `KaleidoscopePlugin` calls a virtual function via a pointer. For the `MouseKeys_` class, our compiler proofs that it is smart. As it can see that the inlined `MouseKeys_::init()` calls  `MouseWrapper.begin()` in replaces the call to `MouseKeys_::init()` with  `MouseWrapper.begin()` - exactly what we would expect for an inlined method.
```
00000612 <PluginHookAdapter<HookLoop>::init()>:
     612:       0e 94 1c 11     call    0x2238  ; 0x2238 <MouseWrapper_::begin()>
     616:       84 e2           ldi     r24, 0x24       ; 36
     618:       96 e0           ldi     r25, 0x06       ; 6
     61a:       0e 94 02 03     call    0x604   ; 0x604 <KaleidoscopePlugin::init()>
     61e:       84 e2           ldi     r24, 0x24       ; 36
     620:       96 e0           ldi     r25, 0x06       ; 6
     622:       0c 94 02 03     jmp     0x604   ; 0x604 <KaleidoscopePlugin::init()>
```
Just for fun, I added `MouseKeys` 32 times. Only `PluginHookAdapter<HookLoop>::init()` does grow.

Reply:
That was a lot of text, with lots of detail that I didn't really follow. How about just answering a yes or no question with one word: is it "yes" or "no"?

Reply:
Please keep in mind that you are not the only one who is reading this. Other people might be interested why the answer is "no".

Reply:
Thank you. Please keep in mind that when a question is asked, the person who asked it — twice — might actually want an answer to their question, not just an arcane diatribe explaining why the answer is the way it is, while leaving that answer a mystery.

Reply:
>Other people might be interested why the answer is "no".

Reply:
Do you not understand that _you didn't actually say that the answer was "no"?_

Reply:
i.e. I am still not sure what the answer is.

Reply:
No

Reply:
Thank you.